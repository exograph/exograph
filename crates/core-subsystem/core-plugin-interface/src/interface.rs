// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::path::PathBuf;

#[cfg(not(target_family = "wasm"))]
use std::{env::current_exe, path::Path};

use crate::core_model_builder::{
    builder::system_builder::BaseModelSystem, error::ModelBuildingError, plugin::SubsystemBuild,
    typechecker::annotation::AnnotationSpec,
};
use crate::core_resolver::plugin::SubsystemResolver;
use crate::error::ModelSerializationError;
use async_trait::async_trait;
use core_model_builder::typechecker::typ::TypecheckedSystem;
use thiserror::Error;

use crate::build_info::SubsystemCheckError;
use exo_env::Environment;

#[async_trait]
pub trait SubsystemBuilder {
    /// Unique string to identify the subsystem by. Should be shared with the corresponding
    /// [SubsystemLoader].
    fn id(&self) -> &'static str;

    /// Subsystem-specific annotations to typecheck during the building phase.
    /// Implementations should provide information about the annotations this plugin supports.
    /// The output is a [Vec], consisting of tuples with the annotation name and what
    /// parameters and targets it supports ([AnnotationSpec]).
    ///
    /// One particular annotation that all plugins should declare (if nothing else)
    /// is the plugin annotation. Plugin annotations are used to mark what subsystem a module
    /// should be handled by.
    ///
    /// For example, in order to typecheck:
    ///
    /// ```exo
    /// @deno("example.ts")
    /// module ExampleModule {
    ///     ...
    /// ```
    ///
    /// [SubsystemBuilder::annotations] should provide:
    ///
    /// ```ignore
    /// fn annotations(&self) -> Vec<(&'static str, AnnotationSpec)> {
    ///     vec![
    ///         ("deno", AnnotationSpec {
    ///             targets: &[AnnotationTarget::Module],
    ///             no_params: false,
    ///             single_params: true,
    ///             mapped_params: None,
    ///         })
    ///     ]
    /// }
    /// ```
    ///
    fn annotations(&self) -> Vec<(&'static str, AnnotationSpec)>;

    /// Build a subsystem's model, producing an [`Option<SubsystemBuild>`].
    ///
    /// - `typechecked_system`: A partially typechecked system. This contains the set of all types
    ///                         that were successfully parsed from the user's model, ranging from `module` types
    ///                         to composite `type`.
    /// - `base_system`: The base model system for Exograph. These are a set of common types that are
    ///                  used by all plugins, like `context`s and primitive types (`Int`, `String`, etc.)
    ///
    /// Return variants:
    ///
    /// - `Ok(Some(SubsystemBuild { .. }))`: The subsystem was built successfully.
    /// - `Ok(None)`: There were no user-declared modules (no build is required).
    /// - `Err(ModelBuildingError { .. })`: The subsystem was not built successfully.
    async fn build(
        &self,
        typechecked_system: &TypecheckedSystem,
        base_system: &BaseModelSystem,
    ) -> Result<Option<SubsystemBuild>, ModelBuildingError>;
}

#[async_trait]
pub trait SubsystemLoader {
    /// Unique string to identify the subsystem by. Should be shared with the corresponding
    /// [SubsystemBuilder].
    fn id(&self) -> &'static str;

    /// Loads and initializes the subsystem, producing a [SubsystemResolver].
    async fn init(
        &mut self,
        serialized_subsystem: Vec<u8>,
        env: &dyn Environment,
    ) -> Result<Box<dyn SubsystemResolver + Send + Sync>, SubsystemLoadingError>;
}

#[derive(Error, Debug)]
pub enum SubsystemLoadingError {
    #[error("System serialization error: {0}")]
    ModelSerializationError(#[from] ModelSerializationError),

    #[error("{0}")]
    BoxedError(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),

    #[error("Configuration error: {0}")]
    Config(String),
}

#[derive(Error, Debug)]
pub enum LibraryLoadingError {
    #[error("Library not found at {0}")]
    LibraryNotFound(PathBuf),

    #[cfg(not(target_family = "wasm"))]
    #[error("Error while loading subsystem library: {0}")]
    LibraryLoadingError(#[from] libloading::Error),

    #[cfg(not(target_family = "wasm"))]
    #[error("Error while loading symbol {0} from library: {0}")]
    SymbolLoadingError(String, libloading::Error),

    #[error("Error while opening subsystem library: {0}")]
    IoError(#[from] std::io::Error),

    #[error("Subsystem library check failed: {0}")]
    CheckFailed(#[from] SubsystemCheckError),
}

#[cfg(not(target_family = "wasm"))]
/// Loads a constructor function from a subsystem library and invokes it.
/// Returns the resultant object.
///
/// * `library_name` - The name of the library to load (platform-independent).
/// * `constructor_symbol_name` - The symbol the constructor function is under in the library.
fn load_subsystem_library<T: ?Sized>(
    library_path: &Path,
    constructor_symbol_name: &str,
) -> Result<Box<T>, LibraryLoadingError> {
    // load the dynamic library
    let lib = Box::new(
        // SAFETY: see documentation for [libloading::Library::new]
        unsafe { libloading::Library::new(library_path.as_os_str())? },
    );

    // check the subsystem's build info and make sure it is valid to load
    crate::build_info::check_subsystem_library(&lib)?;

    // get the constructor's pointer
    // SAFETY: this is safe as long as the constructor function loaded
    //  a. has no arguments and returns a pointer
    //  b. returns a pointer to a boxed instance of T
    // this needs to be manually guaranteed by matching the symbol names and types appropriately
    let boxed_obj: Box<T> = unsafe {
        let constructor: libloading::Symbol<unsafe extern "C" fn() -> *mut T> =
            lib.get(constructor_symbol_name.as_bytes()).map_err(|e| {
                LibraryLoadingError::SymbolLoadingError(constructor_symbol_name.to_string(), e)
            })?;

        let obj_raw = constructor(); // construct the object and get its pointer
        Box::from_raw(obj_raw) // construct from struct pointer
    };

    Box::leak(lib); // keep library alive & never drop

    // return object
    Ok(boxed_obj)
}

#[cfg(not(target_family = "wasm"))]
/// Loads a subsystem builder from a dynamic library.
pub fn load_subsystem_builder(
    library_path: &Path,
) -> Result<Box<dyn SubsystemBuilder + Send + Sync>, LibraryLoadingError> {
    load_subsystem_library(library_path, "__exograph_subsystem_builder")
}

#[cfg(not(target_family = "wasm"))]
/// Loads a subsystem loader from a dynamic library.
pub fn load_subsystem_loader(
    library_name: &str,
) -> Result<Box<dyn SubsystemLoader + Send + Sync>, LibraryLoadingError> {
    // search executable directory for library
    // TODO: we should try to load from sources LD_LIBRARY_PATH first
    let mut library_path = current_exe()?;
    library_path.pop();
    library_path.push(libloading::library_filename(library_name));

    if !library_path.exists() {
        return Err(LibraryLoadingError::LibraryNotFound(library_path));
    }

    load_subsystem_library(&library_path, "__exograph_subsystem_loader")
}
