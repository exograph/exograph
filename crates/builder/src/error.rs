use codemap_diagnostic::Diagnostic;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParserError {
    // Don't include the source, because we emit is as a diagnostic
    #[error("Could not process input clay files")]
    Diagnosis(Vec<Diagnostic>),

    #[error("File '{0}' not found")]
    FileNotFound(String),

    #[error("{0}")]
    IO(#[from] std::io::Error),

    #[error("{0}")]
    ModelBuildingError(#[from] core_model_builder::error::ModelBuildingError),

    #[error("{0}")]
    ModelSerializationError(#[from] core_plugin_shared::error::ModelSerializationError),

    #[error("{0}")]
    Generic(String),
}
