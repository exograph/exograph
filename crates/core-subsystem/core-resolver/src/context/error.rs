// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use thiserror::Error;

#[derive(Debug, Error)]
pub enum ContextParsingError {
    #[error("Could not find source `{0}`")]
    SourceNotFound(String),

    #[error("Unauthorized request")]
    Unauthorized,

    #[error("Malformed request")]
    Malformed,

    #[error("{0}")]
    Delegate(#[from] Box<dyn std::error::Error + Send + Sync>),

    #[error("Type mismatch: Expected `{expected}`, found `{actual}`")]
    TypeMismatch { expected: String, actual: String },

    #[error("Field not found: `{0}`")]
    FieldNotFound(String),
}
