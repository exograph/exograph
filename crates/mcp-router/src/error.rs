#[derive(thiserror::Error, Debug)]
pub enum McpRouterError {
    #[error("Invalid protocol version: {0}")]
    InvalidProtocolVersion(String),
}
