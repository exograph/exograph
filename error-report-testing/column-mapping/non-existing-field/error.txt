error[C000]: Field 'invalid' specified in column mapping does not exist in type 'Member'
  --> src/index.exo:15:5
   |
15 |     @column(mapping={invalid: "membership_member_id", memberTenantId: "membership_tenant_id"}) member: Member?
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Parser error: Could not process input exo files

