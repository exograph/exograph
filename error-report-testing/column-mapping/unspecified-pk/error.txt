error[C000]: Primary key field 'memberId' from type 'Member' is missing in the column mapping
  --> src/index.exo:12:5
   |
12 |     @column(mapping={memberTenantId: "membership_tenant_id"}) member: Member
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Parser error: Could not process input exo files

