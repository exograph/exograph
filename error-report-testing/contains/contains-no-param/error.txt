error[C000]: Contains function expects one parameter
 --> src/index.exo:7:29
  |
7 |   @access(AuthContext.roles.contains())
  |                             ^^^^^^^^^^ type mismatch

Error: Parser error: Could not process input exo files

