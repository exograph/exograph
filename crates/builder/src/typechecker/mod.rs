use std::collections::HashMap;

use codemap::Span;
use codemap_diagnostic::{Diagnostic, Level, SpanLabel, SpanStyle};
use core_model::{mapped_arena::MappedArena, primitive_type::PrimitiveType};
use core_model_builder::{
    ast::ast_types::{AstModel, AstModelKind, AstService, AstSystem, Untyped},
    typechecker::{
        annotation::{AnnotationSpec, AnnotationTarget, MappedAnnotationParamSpec},
        typ::{Service, Type, TypecheckedSystem},
        Scope,
    },
};
use core_plugin_interface::interface::SubsystemBuilder;

use crate::error::ParserError;

mod annotation;
pub mod annotation_map;
mod annotation_params;
mod expression;
mod field;
mod field_default_value;
mod field_type;
mod logical_op;
mod model;
mod relational_op;
mod selection;
mod service;

pub trait TypecheckInto<T> {
    fn shallow(&self) -> T;
    fn pass(
        &self,
        typ: &mut T,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

pub trait TypecheckFrom<T>
where
    Self: Sized,
{
    fn shallow(untyped: &T) -> Self;
    fn pass(
        &mut self,
        type_env: &MappedArena<Type>,
        annotation_env: &HashMap<String, AnnotationSpec>,
        scope: &Scope,
        errors: &mut Vec<Diagnostic>,
    ) -> bool;
}

fn populate_type_env(env: &mut MappedArena<Type>) {
    // TODO: maybe we don't need to do this manually
    env.add("Boolean", Type::Primitive(PrimitiveType::Boolean));
    env.add("Int", Type::Primitive(PrimitiveType::Int));
    env.add("Float", Type::Primitive(PrimitiveType::Float));
    env.add("Decimal", Type::Primitive(PrimitiveType::Decimal));
    env.add("String", Type::Primitive(PrimitiveType::String));
    env.add("LocalTime", Type::Primitive(PrimitiveType::LocalTime));
    env.add(
        "LocalDateTime",
        Type::Primitive(PrimitiveType::LocalDateTime),
    );
    env.add("LocalDate", Type::Primitive(PrimitiveType::LocalDate));
    env.add("Instant", Type::Primitive(PrimitiveType::Instant));
    env.add("Json", Type::Primitive(PrimitiveType::Json));
    env.add("Blob", Type::Primitive(PrimitiveType::Blob));
    env.add("Uuid", Type::Primitive(PrimitiveType::Uuid));

    env.add("Claytip", Type::Primitive(PrimitiveType::Claytip));
    env.add("ClaytipPriv", Type::Primitive(PrimitiveType::ClaytipPriv));

    env.add(
        "Operation",
        Type::Primitive(PrimitiveType::Interception("Operation".to_string())),
    );
}

fn populate_annotation_env(
    subsystem_builders: &[Box<dyn SubsystemBuilder>],
    env: &mut HashMap<String, AnnotationSpec>,
) {
    let mut annotations = vec![
        (
            "access",
            AnnotationSpec {
                targets: &[
                    AnnotationTarget::Type,
                    AnnotationTarget::Field,
                    AnnotationTarget::Method,
                ],
                no_params: false,
                single_params: true,
                mapped_params: Some(&[
                    MappedAnnotationParamSpec {
                        name: "query",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "mutation",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "create",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "update",
                        optional: true,
                    },
                    MappedAnnotationParamSpec {
                        name: "delete",
                        optional: true,
                    },
                ]),
            },
        ),
        (
            "cookie",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "env",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        #[cfg(feature = "test-context")]
        (
            "test",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "before",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "after",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "around",
            AnnotationSpec {
                targets: &[AnnotationTarget::Interceptor],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "header",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "inject",
            AnnotationSpec {
                targets: &[AnnotationTarget::Argument],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "clientIp",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: false,
                mapped_params: None,
            },
        ),
        (
            "jwt",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: true,
                single_params: true,
                mapped_params: None,
            },
        ),
        (
            "query",
            AnnotationSpec {
                targets: &[AnnotationTarget::Field],
                no_params: false,
                single_params: true,
                mapped_params: None,
            },
        ),
    ];

    for builder in subsystem_builders.iter() {
        annotations.extend(builder.annotations())
    }

    for (name, spec) in annotations {
        env.insert(name.to_owned(), spec);
    }
}

pub fn build(
    subsystem_builders: &[Box<dyn SubsystemBuilder>],
    ast_system: AstSystem<Untyped>,
) -> Result<TypecheckedSystem, ParserError> {
    let mut types_arena: MappedArena<Type> = MappedArena::default();
    let mut services_arena: MappedArena<Service> = MappedArena::default();
    let mut annotation_env = HashMap::new();
    populate_type_env(&mut types_arena);
    populate_annotation_env(subsystem_builders, &mut annotation_env);

    validate_no_duplicates(&ast_system.services, |s| &s.name, |s| s.span, "service")?;

    let mut ast_service_types: Vec<AstModel<Untyped>> = vec![];
    for service in ast_system.services.iter() {
        ast_service_types.extend(service.types.clone());
        services_arena.add(&service.name, Service(AstService::shallow(service)));

        validate_service(service)?;
    }

    let ast_types_iter = ast_system.types.iter().chain(ast_service_types.iter());
    let ast_root_types = &ast_system.types;

    for ast_type in ast_types_iter.clone() {
        types_arena.add(
            ast_type.name.as_str(),
            Type::Composite(AstModel::shallow(ast_type)),
        );
    }

    loop {
        let mut did_change = false;
        let init_scope = Scope {
            enclosing_type: None,
        };

        let mut errors = Vec::new();

        for ast_root_type in ast_root_types.iter() {
            let mut typ = types_arena
                .get_by_key(ast_root_type.name.as_str())
                .unwrap()
                .clone();
            if let Type::Composite(c) = &mut typ {
                if c.kind != AstModelKind::Context {
                    errors.push(Diagnostic {
                        level: Level::Error,
                        message: "Models and types are not permitted outside a service".to_string(),
                        code: Some("C000".to_string()),
                        spans: vec![SpanLabel {
                            span: ast_root_type.span,
                            style: SpanStyle::Primary,
                            label: None,
                        }],
                    })
                }
            }
        }

        // Temporary workaround to avoid reporting errors twice (once here, and once in the service pass)
        // TODO: Remove this after fixing https://github.com/payalabs/payas/issues/596
        let mut ignore_errors = Vec::new();

        for ast_type in ast_types_iter.clone() {
            let mut typ = types_arena
                .get_by_key(ast_type.name.as_str())
                .unwrap()
                .clone();
            if let Type::Composite(c) = &mut typ {
                let pass_res = c.pass(
                    &types_arena,
                    &annotation_env,
                    &init_scope,
                    &mut ignore_errors,
                );
                if pass_res {
                    *types_arena.get_by_key_mut(ast_type.name.as_str()).unwrap() = typ;
                    did_change = true;
                }
            } else {
                errors.push(Diagnostic {
                    level: Level::Error,
                    message: format!("Type {} is not a model", typ),
                    code: Some("C000".to_string()),
                    spans: vec![],
                });
            }
        }

        for ast_service in ast_system.services.iter() {
            let mut service = services_arena
                .get_by_key(ast_service.name.as_str())
                .unwrap()
                .clone();
            let Service(s) = &mut service;

            let pass_res = s.pass(&types_arena, &annotation_env, &init_scope, &mut errors);
            if pass_res {
                *services_arena
                    .get_by_key_mut(ast_service.name.as_str())
                    .unwrap() = service;
                did_change = true;
            }
        }

        if !did_change {
            if !errors.is_empty() {
                return Err(ParserError::Diagnosis(errors));
            } else {
                return Ok(TypecheckedSystem {
                    types: types_arena,
                    services: services_arena,
                });
            }
        }
    }
}

fn validate_service(service: &AstService<Untyped>) -> Result<(), ParserError> {
    validate_no_duplicates(
        &service.methods,
        |method| &method.name,
        |method| method.span,
        "operation",
    )?;

    validate_no_duplicates(
        &service.types,
        |model| &model.name,
        |model| model.span,
        "model/type",
    )
}

fn validate_no_duplicates<T>(
    items: &[T],
    get_name: impl Fn(&T) -> &str,
    get_span: impl Fn(&T) -> Span,
    item_kind: &str, // To print as a part of the error message
) -> Result<(), ParserError> {
    let mut items_with_pos = HashMap::new();
    let mut duplicates_with_pos = vec![];

    for item in items.iter() {
        // TODO: Use try_insert when it's stable. This way the first item will always be designated as the existing one
        // Currently, if we have three duplicates, we get diagnostics as (existing: 1, duplicate: 2), (existing: 2, duplicate: 3)
        let existing_item = items_with_pos.insert(get_name(item), get_span(item));

        if let Some(existing_span) = existing_item {
            duplicates_with_pos.push((get_name(item), existing_span, get_span(item)));
        }
    }

    if duplicates_with_pos.is_empty() {
        Ok(())
    } else {
        let diagnostics =
            duplicates_with_pos
                .into_iter()
                .map(|(name, existing_span, duplicate_span)| Diagnostic {
                    level: Level::Error,
                    message: format!("Duplicate {item_kind}: {name}"),
                    code: Some("C000".to_string()),
                    spans: vec![
                        SpanLabel {
                            span: existing_span,
                            style: SpanStyle::Primary,
                            label: Some("first defined here".to_string()),
                        },
                        SpanLabel {
                            span: duplicate_span,
                            style: SpanStyle::Secondary,
                            label: Some("again defined here".to_string()),
                        },
                    ],
                });

        Err(ParserError::Diagnosis(diagnostics.collect()))
    }
}
#[cfg(test)]
pub mod test_support {
    use codemap::CodeMap;

    use super::*;
    use crate::{load_subsystem_builders, parser::parse_str};

    pub fn build(src: &str) -> Result<TypecheckedSystem, ParserError> {
        let mut codemap = CodeMap::new();
        let parsed = parse_str(src, &mut codemap, "input.clay")?;
        let subsystem_builders = load_subsystem_builders().unwrap();
        super::build(&subsystem_builders, parsed)
    }

    pub fn parse_sorted(src: &str) -> Vec<(String, Type)> {
        let checked = build(src).unwrap();

        let mut entries: Vec<_> = checked
            .types
            .keys()
            .map(|key| (key.clone(), checked.types.get_by_key(key).unwrap().clone()))
            .collect();

        entries.sort_by_key(|pair| pair.0.to_owned());
        entries
    }
}

#[cfg(test)]
mod tests {
    use super::test_support::{build, parse_sorted};

    // Due to a change in insta version 1.12, test names (hence the snapshot names) get derived
    // from the surrounding function, so we must use a macro instead of a helper function.
    macro_rules! assert_typechecking {
        ($src:expr) => {
            insta::with_settings!({sort_maps => true}, {
                insta::assert_yaml_snapshot!(build($src).unwrap())
            });
        };
    }

    #[test]
    fn simple() {
        let src = r#"
        @postgres
        service UserService {
            type User {
              @column("custom_column") @access(self.role == "role_admin" || self.role == "role_superuser" || self.doc.is_public)
              doc: Doc;
              role: String
            }

            type Doc {
              is_public: Boolean
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_auth_context_use_in_field_annotation() {
        let src = r#"
        context AuthContext {
            @jwt role: String
        }

        @postgres
        service DocumentService {
            type Doc {
              is_public: Boolean
              @access(AuthContext.role == "ROLE_ADMIN" || self.is_public) content: String 
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_array_in_operator() {
        let src = r#"
        context AuthContext {
            @jwt roles: Array<String> 
        }

        @postgres
        service DocumentService {
            type Doc {
              @access("ROLE_ADMIN" in AuthContext.roles) content: String 
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn with_auth_context_use_in_type_annotation() {
        let src = r#"
        context AuthContext {
            @jwt role: String
        }
        
        @postgres
        service DocumentService {
            @access(AuthContext.role == "ROLE_ADMIN" || self.is_public)
            type Doc {
              is_public: Boolean
              content: String
            }
        }
        "#;

        assert_typechecking!(src);
    }

    #[test]
    fn insignificant_whitespace() {
        let typical = r#"
        @postgres
        service DocumentService {
            @table("venues")
            type Venue {
                @column("idx") @pk id: Int 
                name: String
            }
        }
        "#;

        let with_whitespace = r#"
        @postgres
        service      DocumentService{
        @table ( "venues" )
        type    Venue
        {
            @column(  "idx"  )    @pk  id:   Int   
           

            name:String

        }}

        "#;

        let typical_parsed = serde_json::to_string(&parse_sorted(typical)).unwrap();
        let with_whitespace_parsed = serde_json::to_string(&parse_sorted(with_whitespace)).unwrap();
        assert_eq!(typical_parsed, with_whitespace_parsed);
    }

    #[test]
    fn unknown_annotation() {
        let src = r#"
        @postgres
        service UserService {
            @asdf
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn duplicate_annotation() {
        let src = r#"
        @postgres
        service UserService {
            @table("users")
            @table("users")
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn duplicate_plugin_annotations() {
        let src = r#"
        @postgres
        @postgres
        service UserService {
            @table("users")
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn no_plugin_annotation() {
        let src = r#"
        service UserService {
            type User {
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn type_at_root() {
        let src_model = r#"
        type User {
        }
        "#;

        assert_err(src_model);
    }

    #[test]
    fn context_in_a_service() {
        let src_model = r#"
        @postgres
        service UserService {
            context AuthContext {
                @jwt role: String
            }
        }
        "#;

        assert_err(src_model);
    }

    #[test]
    fn invalid_annotation_parameter_type() {
        let expected_none = r#"
        @postgres
        service UserService {
            type User {
                @pk("asdf") id: Int 
            }
        }
        "#;

        let expected_single = r#"
        @postgres
        service UserService {
            @table
            type User {
            }
        }
        "#;

        let expected_map = r#"
        @postgres
        service UserService {
            type User {
                @range(5) id: Int
            }
        }
        "#;

        assert_err(expected_none);
        assert_err(expected_single);
        assert_err(expected_map);
    }

    #[test]
    fn duplicate_annotation_mapped_param() {
        let src = r#"
        @postgres
        service UserService {
            type User {
                @range(min=5, max=10, min=3) id: Int
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn unknown_annotation_mapped_param() {
        let src = r#"
        @postgres
        service UserService {
            type User {
                @range(min=5, maxx=10) id: Int
            }
        }
        "#;

        assert_err(src);
    }

    #[test]
    fn invalid_annotation_target() {
        let model = r#"
        @postgres
        service UserService {
            @pk
            type User {
            }
        }
        "#;

        let field = r#"
        @postgres
        service UserService {
            type User {
                @table("asdf") id: Int
            }
        }
        "#;

        assert_err(model);
        assert_err(field);
    }

    #[test]
    fn multiple_types() {
        let model = r#"
        @deno("test.ts")
        service User {
            type User {
                id: Int
            }

            type User {
                id: Int
            }

            query userName(id: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_services() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            query userName(id: Int): String
        }

        @deno("foo.js")
        service Foo {
            query userName(id: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_operations() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            query userName(id1: Int): String
            query userName(id2: Int): String
            query userName(id3: Int): String
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_typess() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_types() {
        let model = r#"
        @postgres
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    #[test]
    fn multiple_same_named_model_and_types() {
        let model = r#"
        @deno("foo.js")
        service Foo {
            type User {
                id: Int
                name: String
            }
            type User {
                id: Int
                name: String
            }
        }
        "#;

        assert_err(model);
    }

    fn assert_err(src: &str) {
        assert!(build(src).is_err());
    }
}
