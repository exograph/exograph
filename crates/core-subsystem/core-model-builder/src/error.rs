use codemap_diagnostic::Diagnostic;
use core_plugin::error::ModelSerializationError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ModelBuildingError {
    #[error("Could not process input clay files")]
    Diagnosis(Vec<Diagnostic>),

    #[error("{0}")]
    IO(#[from] std::io::Error),

    #[error("{0}")]
    Generic(String),

    #[error("Unable to serialize model {0}")]
    Serialize(#[source] ModelSerializationError),

    #[error("Unable to deserialize model {0}")]
    Deserialize(#[source] ModelSerializationError),
}
