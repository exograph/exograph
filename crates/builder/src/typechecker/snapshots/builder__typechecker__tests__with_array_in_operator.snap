---
source: crates/builder/src/typechecker/mod.rs
expression: build(src).unwrap()
---
types:
  values:
    - - ~
      - Primitive: Boolean
    - - ~
      - Primitive: Int
    - - ~
      - Primitive: Float
    - - ~
      - Primitive: Decimal
    - - ~
      - Primitive: String
    - - ~
      - Primitive: LocalTime
    - - ~
      - Primitive: LocalDateTime
    - - ~
      - Primitive: LocalDate
    - - ~
      - Primitive: Instant
    - - ~
      - Primitive: Json
    - - ~
      - Primitive: Blob
    - - ~
      - Primitive: Uuid
    - - ~
      - Primitive: Exograph
    - - ~
      - Primitive: ExographPriv
    - - ~
      - Primitive:
          Interception: Operation
    - - ~
      - Composite:
          name: AuthContext
          kind: Context
          fields:
            - name: roles
              typ:
                Plain:
                  - Array
                  - - Plain:
                        - String
                        - []
                        - true
                  - true
              annotations:
                annotations:
                  jwt:
                    name: jwt
                    params: None
              default_value: ~
          annotations:
            annotations: {}
    - - ~
      - Composite:
          name: Doc
          kind: Type
          fields:
            - name: content
              typ:
                Plain:
                  - String
                  - []
                  - true
              annotations:
                annotations:
                  access:
                    name: access
                    params:
                      Single:
                        - RelationalOp:
                            In:
                              - StringLiteral:
                                  - ROLE_ADMIN
                              - FieldSelection:
                                  Select:
                                    - Single:
                                        - Identifier:
                                            - AuthContext
                                            - Reference:
                                                index: 15
                                                generation: ~
                                        - Reference:
                                            index: 15
                                            generation: ~
                                    - Identifier:
                                        - roles
                                        - Defer
                                    - Array:
                                        Reference:
                                          index: 4
                                          generation: ~
                              - Primitive: Boolean
              default_value: ~
          annotations:
            annotations: {}
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
  map:
    AuthContext:
      index: 15
      generation: ~
    Blob:
      index: 10
      generation: ~
    Boolean:
      index: 0
      generation: ~
    Decimal:
      index: 3
      generation: ~
    Doc:
      index: 16
      generation: ~
    Exograph:
      index: 12
      generation: ~
    ExographPriv:
      index: 13
      generation: ~
    Float:
      index: 2
      generation: ~
    Instant:
      index: 8
      generation: ~
    Int:
      index: 1
      generation: ~
    Json:
      index: 9
      generation: ~
    LocalDate:
      index: 7
      generation: ~
    LocalDateTime:
      index: 6
      generation: ~
    LocalTime:
      index: 5
      generation: ~
    Operation:
      index: 14
      generation: ~
    String:
      index: 4
      generation: ~
    Uuid:
      index: 11
      generation: ~
modules:
  values:
    - - ~
      - name: DocumentModule
        annotations:
          annotations:
            postgres:
              name: postgres
              params: None
        types:
          - name: Doc
            kind: Type
            fields:
              - name: content
                typ:
                  Plain:
                    - String
                    - []
                    - true
                annotations:
                  annotations:
                    access:
                      name: access
                      params:
                        Single:
                          - RelationalOp:
                              In:
                                - StringLiteral:
                                    - ROLE_ADMIN
                                - FieldSelection:
                                    Select:
                                      - Single:
                                          - Identifier:
                                              - AuthContext
                                              - Reference:
                                                  index: 15
                                                  generation: ~
                                          - Reference:
                                              index: 15
                                              generation: ~
                                      - Identifier:
                                          - roles
                                          - Defer
                                      - Array:
                                          Reference:
                                            index: 4
                                            generation: ~
                                - Primitive: Boolean
                default_value: ~
            annotations:
              annotations: {}
        methods: []
        interceptors: []
        base_exofile: input.exo
    - ~
    - ~
    - ~
  map:
    DocumentModule:
      index: 0
      generation: ~

